geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/writer/tx/AccumuloAtomicIndexWriter.scala

```diff
@@ -0,0 +1,208 @@
+          // any previous values that are overwritten are added as conditions on the put
+          // any previous values that aren't overwritten need to be deleted
+          val remaining = toUpdate.kvs.filterNot(kv => mutations.kvs.exists(_.equalKey(kv)))
+          if (remaining.nonEmpty) {
+            dels.append(toUpdate.copy(kvs = remaining))
+          }
+          UpdateBuilder(mutations.row, mutations.kvs, toUpdate.kvs)
+        }
+      }
+      // note: order here is important, we need puts to operate before deletes in order for conditions to succeed
+      mutations(i) = puts ++ dels
+      i += 1
+    }
+    mutate(feature.feature.getID, mutations)
+  }
+
+  /**
+   * Apply the mutations
+   *
+   * @param id feature id
+   * @param mutations mutations to apply
+   */
+  @throws[ConditionalWriteException]
+  private def mutate(id: String, mutations: Array[Seq[MutationBuilder]]): Unit = {
+    val errors = ArrayBuffer.empty[ConditionalWriteStatus]
+    val successes = ArrayBuffer.empty[(Int, MutationBuilder)]
+    var i = 0
+    while (i < mutations.length) {
+      // note: mutations need to be applied in order so that conditions are correct
+      mutations(i).foreach { m =>
+        val status = writers(i).write(m.apply()).getStatus
+        if (status == ConditionalWriter.Status.ACCEPTED ||
+            (status == ConditionalWriter.Status.UNKNOWN && verifyWrite(tables(i), m))) {
+          successes += i -> m
+        } else {
+          errors += ConditionalWriteStatus(indices(i), m.name, status)
+        }
+      }
+      i += 1
+    }
+    if (errors.nonEmpty) {
+      // revert any writes we made
+      successes.foreach { case (i, m) =>
+        // note: if these are rejected, some other update has come through and we don't want to overwrite it
+        writers(i).write(m.invert())
+      }
+      throw ConditionalWriteException(id, errors.toSeq)
+    }
+  }
+
+  /**
+   * Verify if a mutation was successfully applied or not
+   *
+   * @param table table being mutated
+   * @param mutation mutation being applied
+   * @return true if the current state of the table reflects the mutation's intended result
+   */
+  private def verifyWrite(table: String, mutation: MutationBuilder): Boolean = {
+    WithClose(new IsolatedScanner(ds.connector.createScanner(table, ds.auths))) { scanner =>
+      val key = new Key(mutation.row)
+      scanner.setRange(new org.apache.accumulo.core.data.Range(key, true, key.followingKey(PartialKey.ROW), false))
+      val found = scanner.iterator().asScala.toList
+      mutation.apply().getUpdates.asScala.forall { update =>
+        if (update.isDeleted) {
+          found.forall { entry =>
+            !compare(entry.getKey.getColumnFamily, update.getColumnFamily) ||
+                !compare(entry.getKey.getColumnQualifier, update.getColumnQualifier) ||
+                !compare(entry.getKey.getColumnVisibility, update.getColumnVisibility) ||
+                entry.getValue.getSize > 0
```

Comment:
Why do we care that the found entry has a value?

Reply:
i think this check is backwards actually... it's checking to see if a row was deleted successfully, so if it has a value then it wasn't deleted (deletes have empty values, although i don't think they should actually come back from a scan anyway...). does changing it to `entry.getValue.getSize == 0` make more sense?

Reply:
```suggestion
                entry.getValue.getSize == 0
```

Reply:
I think I was basing this on the `putDelete` method in Accumulo, which sets an empty byte array: https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/data/Mutation.java#L423-L425

Actually the key has a deleted check, so I should probably just look at that:

```suggestion
                entry.getKey.isDeleted
```